feat: Add Node 22 support

mtrezza opened the pull request:
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Approach

Just trying out CI; this also removes specific Node.js versions for testing; still to be discussed whether we want that.

## Tasks
<!-- Delete tasks that don't apply. -->

mtrezza committed:
add node 22

mtrezza committed:
replace punycode

mtrezza committed:
fit

dplewis commented on spec/CLI.spec.js:
```diff
@@ -220,7 +220,7 @@ describe('execution', () => {
   it('should start Parse Server', done => {
     const env = { ...process.env };
-    env.NODE_OPTIONS = '--dns-result-order=ipv4first';
+    // env.NODE_OPTIONS = '--dns-result-order=ipv4first';
```
Isn't this handled by
https://github.com/parse-community/parse-server/blob/alpha/spec/helper.js#L10

mtrezza replied:
Yes, that may not be needed anymore, I disabled it to try without

mtrezza replied:
The tests fail when removing it, I think that is because the tests are spawning up a new process, hence this env var needs to be applied before that.

mtrezza commented:
@dplewis the issue is that there are 4 tests failing. The [punycode library](https://github.com/mathiasbynens/punycode.js) that should be used instead of the deprecated Node.js feature describes how to implement it, but that doesn't seem to work. It still seems to be using the Node.js feature.

mtrezza committed:
remove ipv4

mtrezza committed:
Revert "remove ipv4"

This reverts commit 4e947bcb85c99fe0e7c90012cf72b1fa4ca6b3fe.

mtrezza committed:
re-add ipv4

mtrezza committed:
Update mongodbUrl.js

mtrezza committed:
Merge branch 'alpha' into feat/node-22

mtrezza committed:
Update CLI.spec.js

mtrezza commented:
From the failing tests I assume this these deprecation warnings come from certain dependencies which have a sub-dependency on Node.js internal punycode. With `--trace-deprecation` enabled the trace is:

```
1) execution should start Parse Server
  - Failed: (node:2521) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
      at node:punycode:3:9
      at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/realm:398:7)
      at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/realm:337:10)
      at loadBuiltinModule (node:internal/modules/helpers:96:7)
      at Module._load (node:internal/modules/cjs/loader:1070:17)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
      at Module.require (node:internal/modules/cjs/loader:1304:12)
      at require (node:internal/modules/helpers:123:16)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/tr46/index.js:3:18)
      at Module._compile (node:internal/modules/cjs/loader:1467:14)
      at Module.replacementCompile (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:60:13)
      at Module._extensions..js (node:internal/modules/cjs/loader:1551:10)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:64:4)
      at Module.load (node:internal/modules/cjs/loader:1282:32)
      at Module._load (node:internal/modules/cjs/loader:1098:12)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
      at Module.require (node:internal/modules/cjs/loader:1304:12)
      at require (node:internal/modules/helpers:123:16)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/whatwg-url/lib/url-state-machine.js:2:14)
      at Module._compile (node:internal/modules/cjs/loader:1467:14)
      at Module.replacementCompile (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:60:13)
      at Module._extensions..js (node:internal/modules/cjs/loader:1551:10)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:64:4)
      at Module.load (node:internal/modules/cjs/loader:1282:32)
      at Module._load (node:internal/modules/cjs/loader:1098:12)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
      at Module.require (node:internal/modules/cjs/loader:1304:12)
      at require (node:internal/modules/helpers:123:16)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/whatwg-url/lib/URL-impl.js:2:13)
      at Module._compile (node:internal/modules/cjs/loader:1467:14)
      at Module.replacementCompile (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:60:13)
      at Module._extensions..js (node:internal/modules/cjs/loader:1551:10)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:64:4)
      at Module.load (node:internal/modules/cjs/loader:1282:32)
      at Module._load (node:internal/modules/cjs/loader:1098:12)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
      at wrapModuleLoad (node:internal/modules/cjs/loader:215:24)
      at Module.require (node:internal/modules/cjs/loader:1304:12)
      at require (node:internal/modules/helpers:123:16)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/whatwg-url/lib/URL.js:442:14)
      at Module._compile (node:internal/modules/cjs/loader:1467:14)
      at Module.replacementCompile (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:60:13)
      at Module._extensions..js (node:internal/modules/cjs/loader:1551:10)
      at Object.<anonymous> (/home/runner/work/parse-server/parse-server/node_modules/append-transform/index.js:64:4)
      at Module.load (node:internal/modules/cjs/loader:1282:32)
      at Module._load (node:internal/modules/cjs/loader:1098:12)
      at TracingChannel.traceSync (node:diagnostics_channel:315:14)
```

I believe these errors come from 3rd party dependencies, so we cannot do much about it. A solution is to change the tests so they ignore deprecation warnings in the stderr output. This is just a deprecation warnings after all, not an error. The history of [DEP0040](https://nodejs.org/dist/latest-v21.x/docs/api/deprecations.html#DEP0040) shows that while the deprecation happened in Node 7, only since Node 21 is it a runtime deprecation, so that's why it became visible now.

mtrezza committed:
remove punycode

mtrezza committed:
Update CLI.spec.js

mtrezza committed:
rewrite

mtrezza committed:
Update CLI.spec.js

mtrezza commented:
Changed the tests so that deprecation gets ignored; once the feature gets removed from Node.js, the process should fail to launch a Parse Server instance and the tests will also fail.

Also removed Parse Server's internal code reliance on Node.js punycode feature to make it ready for the feature removal.

mtrezza committed:
add punycode

parseplatformorg commented:
🎉 This change has been released in version [7.3.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/7.3.0-alpha.1)

parseplatformorg commented:
🎉 This change has been released in version [7.3.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/7.3.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [7.3.0](https://github.com/parse-community/parse-server/releases/tag/7.3.0)
